pkg/sql/conn_io.go

```diff
@@ -457,16 +457,17 @@ func (buf *StmtBuf) translatePosLocked(pos CmdPos) (int, error) {
 	return int(pos - buf.mu.startPos), nil
 }
 
-// ltrim iterates over the buffer forward and removes all commands up to
+// Ltrim iterates over the buffer forward and removes all commands up to
 // (not including) the command at pos.
 //
-// It's illegal to ltrim to a position higher than the current cursor.
-func (buf *StmtBuf) ltrim(ctx context.Context, pos CmdPos) {
+// It's illegal to Ltrim to a position higher than the current cursor.
+func (buf *StmtBuf) Ltrim(ctx context.Context, pos CmdPos) {
 	buf.mu.Lock()
 	defer buf.mu.Unlock()
 	if pos < buf.mu.startPos {
-		log.Fatalf(ctx, "invalid ltrim position: %d. buf starting at: %d",
-			pos, buf.mu.startPos)
+		// already trimmed, this can happen when the StmtBuf is trimmed from the limitedCommandResult
+		// side state machine and then again in the conn_executor.
```

Comment:
at what point after using a MaxRows portal does conn_executor again try to Ltrim the buffer? i'm wondering if it's possible to teach conn_executor not to do that, so we can keep this error and maybe be a little safer.

Reply:
After the portal is exhausted in `limitedCommandResult` the flow returns to [here](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/conn_executor.go#L1671). `Ltrim` is then called again at the bottom of the function [here](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/conn_executor.go#L1813). At this point conn_executor  uses `pos` to `Ltrim` but that index has already been trimmed. I considered updating `pos` after `execPortal` returns with something like `cmd, pos, err := ex.stmtBuf.CurCmd()`, which would give the cleanup routine in conn_executor a valid index to `LTrim`, but wasn't sure if I was missing some unwanted side effects in the rewind determination code. Let me know if you think that is a better approach.

Reply:
Made the change to update the state in conn_executor when execPortal returns.